core/base/scalarFieldCriticalPoints/ScalarFieldCriticalPoints.h

```diff
@@ -88,18 +88,27 @@ namespace ttk {
     void checkProgressivityRequirement(const triangulationType *triangulation);
 
     template <class triangulationType = AbstractTriangulation>
-    int getNumberOfLowerUpperComponents(const SimplexId vertexId,
-                                        const SimplexId *const offsets,
-                                        const triangulationType *triangulation,
-                                        ttk::SimplexId &lowerComponentNumber,
-                                        ttk::SimplexId &upperComponentNumber,
-                                        bool &isLowerOnBoundary,
-                                        bool &isUpperOnBoundary) const;
+    int getNumberOfLowerUpperComponents(
```

Comment:
I'd suggest to re-model this function as follows:
- rename the function into `getLowerAndUpperComponents(`
- remove the parameters `ttk::SimplexId &lowerComponentNumber,` and `ttk::SimplexId &upperComponentNumber,` (in the using call, we'll simply call the size of the components, see below)
- force the usage of a reference of vectors for the output (without default value to `nullptr`): `std::vector<std::vector<ttk::SimplexId>> &upperComponents` and `std::vector<std::vector<ttk::SimplexId>> &lowerComponents`
- actually compute the lower and upper components even in case of extrema (this is easy, before lines 417).